### ISSUE TITLE
ClearType rendering issue with transformed components (was "Braces are asymmetric for some font sizes")

### ISSUE THREAD

**Issue by ForNeVeR** [rsms/inter#251]
For some font sizes (e.g. 16px) braces (`[](){}`) look asymmetric, the right one may be drawn lower or higher than the left one.

For example, visit this link: https://rsms.me/inter/lab/?size=16&antialias=default and enter braces into the first text line:

**Expected behavior**
Braces should be symmetric and have the same height/position for all font sizes.

**Screenshots**
![image](https://user-images.githubusercontent.com/92793/77726112-bfad5800-7029-11ea-863c-840f992c429b.png)

**Environment**
- OS: Windows 10 1909
- App that renders the font: Chrome
- Display size/scale: 3840 × 2160, 150%
- Version of font: sorry, I don't know; wasn't able to find this information on the font test site

**Comment by rsms** [comment]
Interesting. This seems to be a Windows-specific issue as I an unable to reproduce in Chrome, Safari or Firefox on mac. I will check on my Windows computer as well, but in the meantime, would you mind trying the following things?
- Check in a different browser like Edge (stable, non-chromium; the one that ships with Windows) and/or Firefox? Do you see the same issue there?
- Turn on variable font ([use this link for a shortcut](http://localhost:3003/?size=30&sample=ABCDEFGHIJKLMNOPQRSTUVWXYZ%5B%5D%7B%7D()&varfont=1))

**Comment by originecode** [comment]
I get the same issue on : 
- **OS** : macOS Catalina 10.15.4
- **Inter Version** : Version 3.012;git-f93a4a705
- **Browsers** :
___
- Safari : Version 13.1 (15609.1.20.111.8)
<img width="254" alt="Screenshot 2020-04-04 at 23 00 40" src="https://user-images.githubusercontent.com/42325924/78461423-102b5200-76c9-11ea-9b38-1d51db7bf898.png">

___

- Safari Technology Preview : Release 103 (Safari 13.2, WebKit 15610.1.7.3.2)
<img width="253" alt="Screenshot 2020-04-04 at 23 00 05" src="https://user-images.githubusercontent.com/42325924/78461445-47016800-76c9-11ea-9de0-21a936d40324.png">

___

- Firefox Developer Edition : 75.0b11 & 75.0b12
<img width="258" alt="Screenshot 2020-04-04 at 23 05 14" src="https://user-images.githubusercontent.com/42325924/78461453-541e5700-76c9-11ea-94b0-c492f897f136.png">

**Comment by be5invis** [comment]
@rsms Should be another issue related with transformed component references. This is how GID 1331 shown in VTT:
![image](https://user-images.githubusercontent.com/240091/78468487-6bf7da00-76cd-11ea-9f92-3f412f251d22.png)
Not this component is also set rounded, so Macintosh may also have problems.

@schriftgestalt is it possible to tell Glyphs **not** to use transformed components? Font tools really really dislike that. Glyphs could unlink a component if it is trasnformed.

**Comment by Felix-MFU** [comment]
The experienced the same issue on Sketch (MacOS Catalina). It seems to have to do with the calt (Contextual Alternates feature. It works on capital letters only after letter, but not before it. So the left brace gets positioned wrongly (no reposition), and the right one correctly (reposition).

In Chrome on the other hand braces are positioned correctly.

**Comment by schriftgestalt** [comment]
Sorry for taking so long. Flipped or rotated components are decomposed by default.

**Comment by be5invis** [comment]
@schriftgestalt It looks like Inter used its own script to generate TTF from Glyphs source 🤦. Time to fix that.

**Comment by rsms** [comment]
Indeed this looks like an issue with intended behavior with unexpected results — i.e. not a bug but also not ideal. Inter uses an OpenType feature called "contextual alternates" which is used to replace certain glyphs in certain contexts with specialized glyphs. For example, Inter provides a left parenthesis specialized for capital letters. When a left parenthesis is adjacent to a capital letter, it is replaced by the special one. However, the rules are more complex that this and includes exceptions, which is what causes the issue you see. Specifically, "x (M" vs "(M" — in the first case the adjacent lower-case letter "x" invokes an exception to not perform the substitution.

**Comment by rsms** [comment]
I made some improvements in https://github.com/rsms/inter/commit/ebdfaf21cf499b685774e6be776e6596d81cc272

Here is the result of `/lab/?sample=Feature%3A%20calt&size=24&varWeightNum=446` (master, not the website)
<img width="1965" alt="Screen Shot 2020-08-17 at 17 15 55" src="https://user-images.githubusercontent.com/47046/90456272-5e0b8a00-e0ad-11ea-801c-f7ca437fb466.png">

**Comment by rsms** [comment]
Some improvements to this has been released as part of [version 3.14](https://github.com/rsms/inter/releases/tag/v3.14)

**Comment by ForNeVeR** [comment]
I'm sorry, but I can't see any improvements.

I can see that the braces are rendered differently if there's an adjacent capital letter (they go up a little), but the opening and closing brace still have different lifting, which, on my screen, differs from what I see on your screenshot. Here're the current results from https://rsms.me/inter/lab/?antialias=default&size=16 (which, I believe, has been updated to use 3.14 already? By the way, could we have a font version being printed on the website?)
![image](https://user-images.githubusercontent.com/92793/90490966-7e008380-e169-11ea-8de8-19cf6e60bb7d.png)

And, for comparison, with space or a non-capital letter before and of the braces:
![image](https://user-images.githubusercontent.com/92793/90491036-953f7100-e169-11ea-884e-e128a4842af7.png)

(they're drawn lower, but opening and closing brace are still noticeably different)

**Comment by rsms** [comment]
Interesting! So perhaps this is Windows-specific after all. I’ll test on a Windows machine...

**Comment by be5invis** [comment]
@rsms You can try with Visual TrueType, select the glyphs, select an anti-aliasing mode (TTFA should using "CT(anti-aliased)", enable grid fitting (`Ctrl+G`) and zoom in to see the pixel-level difference.

**Comment by rsms** [comment]
Thanks! I'll try that.

**Comment by be5invis** [comment]
This issue may also influence certain Linux users (depending their `fontconfig` settings).

**Comment by rsms** [comment]
I've spend a good deal of time this evening trying to reproduce the issue but I'm falling flat. I've tried all major browser, macOS and Windows 10 as well as following @be5invis's advice with Visual TrueType. The results I'm seeing are the expected ones (i.e. "()" are aligned and not offset.) Is there any details we might be missing here?

How I'm testing:
1. Open `https://rsms.me/inter/lab/` in a web browser
2. Select all and paste to replace the sample below
3. Inspect and screenshot

sample:
```
ABCDEFGHIJKLMNOPQRSTUVWXYZ[]{}()
abcdefghijklmnopqrstuvwxyz[]{}()
0123456789!?.
```

Note that `https://rsms.me/inter/lab/` does _not_ use font files with TTF hints.

Some screenshots of what I'm seeing:

Visual TrueType, setup as suggested in comment https://github.com/rsms/inter/issues/251#issuecomment-675746560
<img width="764" alt="Screen Shot 2020-08-18 at 18 37 33" src="https://user-images.githubusercontent.com/47046/90582372-5d90f300-e182-11ea-95ad-5191558ab1e3.png">

---
macOS:
<img width="462" alt="Screen Shot 2020-08-18 at 18 25 40" src="https://user-images.githubusercontent.com/47046/90582379-61bd1080-e182-11ea-851a-543186ce914e.png">
<img width="455" alt="Screen Shot 2020-08-18 at 18 25 54" src="https://user-images.githubusercontent.com/47046/90582383-67b2f180-e182-11ea-8541-aee1f3780f9b.png">
<img width="464" alt="Screen Shot 2020-08-18 at 18 26 19" src="https://user-images.githubusercontent.com/47046/90582384-6a154b80-e182-11ea-97f8-287a44c993de.png">

---
Windows 10:
<img width="465" alt="Screen Shot 2020-08-18 at 18 41 39" src="https://user-images.githubusercontent.com/47046/90582465-a34dbb80-e182-11ea-8b73-e0cf2a61ef84.png">

**Comment by be5invis** [comment]
@rsms You could try using ↑ or ↓ to increase or decrease PPEM in VTT. This issue **only occurs** at certain PPEM sizes.

**Comment by rsms** [comment]
Woah. This is so crazy. You're right: at certain sizes the right parenthesis is drawn with an offset on the Y axis. I wonder if this is a bug in the Windows font renderer? (since I can't reproduce this in any web browser on macOS, at any rasterization size.)

<img width="1740" alt="Screen Shot 2020-08-18 at 19 20 01" src="https://user-images.githubusercontent.com/47046/90584959-97fd8e80-e188-11ea-8c26-3b73448c95cb.png">

It's likely so that the component transformation gets messed up somehow on Windows.

<img width="1062" alt="Screen Shot 2020-08-18 at 19 22 24" src="https://user-images.githubusercontent.com/47046/90584986-aba8f500-e188-11ea-9db3-2ec6743a8ba1.png">

Tomorrow I'll try to decompose everything in the font (which will explore the file size but allow me to test this theory.) If this is the case, with component transformation issues on Windows, then perhaps I could offer a decomposed and larger font for Windows specifically.

**Comment by be5invis** [comment]
@rsms It is a feature: you are using rounded component references in `)`, **and** this component is flipped, **and** you have hinting, so the rasterizer correctly interpreted the hinted `(`, rotated it, and translated but using an rounded offset, so they become asymmetric, since you added a lot of rounding (component offset **and** hinting).

**Comment by rsms** [comment]
Hm, but there are no hints in these fonts. The hinted TTF fonts are not served from the website (`/inter/lab/`).

My understanding of the issue, as I read it in the description of this issue, is that things are rendered incorrectly in the web browser when viewing `https://rsms.me/inter/lab/`, which does not serve font files with hints.

Let me ask a clarifying question: Is it possible that we are discussing two different issues here? I.e. rendering in web browsers AND hinting data of the hinted fonts.

What I'm testing and debugging is the rendering of `https://rsms.me/inter/lab/` (as described in the OP.)

This is what I'm seeing walking through rasterization sizes in Visual TrueType, using the most recent release: (v3.14, no hints. two videos stitched together)
![dfs8n91iukjs](https://user-images.githubusercontent.com/47046/90585756-71d8ee00-e18a-11ea-8363-715015d11002.gif)

**Comment by be5invis** [comment]
@rsms Could you press `CTRL + 2` and show me the glyph program of `)`?

**Comment by be5invis** [comment]
`SOFFSET[r]` + transformed components, perfectly symmetric:
![image](https://user-images.githubusercontent.com/240091/90587079-69ce7d80-e18d-11ea-8345-7edddbf700b4.png)

`SOFFSET[R]` + transformed components, slightly asymmetric:
![image](https://user-images.githubusercontent.com/240091/90587099-7521a900-e18d-11ea-942e-e2ed2c56251b.png)

The best solution to this should be unlinking all the components that is transformed -- which shouldn't have a lot.

**Comment by schriftgestalt** [comment]
You can skip all tests on MacOS as it ignores TrueType instructions.

**Comment by rsms** [comment]
@be5invis This is what the CTRL+2 shows me for the glyph program
<img width="968" alt="Screen Shot 2020-08-19 at 09 43 48" src="https://user-images.githubusercontent.com/47046/90665235-a25b6f00-e200-11ea-8838-9a0e34e59d83.png">

**Comment by rsms** [comment]
Now, if I open a font file _with TTF hints_ (from ttfautohint) I do see a program for the glyph as expected, but as I mentioned earlier, the fonts on the website do not have TTF hints.
<img width="955" alt="Screen Shot 2020-08-19 at 09 50 55" src="https://user-images.githubusercontent.com/47046/90666013-a936b180-e201-11ea-8583-0a5f8d10d147.png">

**Comment by rsms** [comment]
Repro on Windows with ClearType enabled: (or Linux with ClearType or similar technology enabled in e.g. FreeType)
1. Install or load Inter Regular with TrueType hints: [Inter-Regular-hinted-v3.14.ttf.zip](https://github.com/rsms/inter/files/5098862/Inter-Regular-hinted-v3.14.ttf.zip) (release v3.14)
2. Use an app that uses the system font renderer, like Notepad
3. Type "ABC()[]{}"
4. Set font to Inter Regular, size 12

This should yield right parenthesis with an unwanted offset on the y axis, as shown in this screenshot:

<img width="859" alt="Screen Shot 2020-08-19 at 12 39 06" src="https://user-images.githubusercontent.com/47046/90681905-0427d300-e219-11ea-9997-22faf24c3336.png">

**Comment by rsms** [comment]
Here's a new build of the fonts, incorporating the changes landed in 7d70535c01803b9c734147acb3049a74875f6356
[Inter-3.14-text-ca7d79eb20-issue-251.zip](https://github.com/rsms/inter/files/5099256/Inter-3.14-text-ca7d79eb20-issue-251.zip)

This is what it looks like on Windows 10 in Notepad:
<img width="1363" alt="Screen Shot 2020-08-19 at 13 52 25" src="https://user-images.githubusercontent.com/47046/90688829-e0b65580-e223-11ea-81b2-d80403d1eccc.png">

This is what is looks like on the website in Windows 10: (left is static font, right is variable)
<img width="838" alt="Screen Shot 2020-08-19 at 14 04 08" src="https://user-images.githubusercontent.com/47046/90689434-f0826980-e224-11ea-98f9-8c3047199a62.png">

**Comment by rsms** [comment]
@ForNeVeR @aemi-dev Would you mind installing the fonts in the zip attached in the [previous comment above](https://github.com/rsms/inter/issues/251#issuecomment-676701245) and let me know if that solves the issue for you?

**Comment by aemi-dev** [comment]
Issue is solved. I'm using macOS.
However :

*Firefox*
<img width="375" alt="ffox" src="https://user-images.githubusercontent.com/42325924/90692524-91007a00-e275-11ea-87d3-4524370b39c7.png">
*Google Chrome*
<img width="375" alt="gchrome" src="https://user-images.githubusercontent.com/42325924/90692527-9231a700-e275-11ea-9d06-20b8e8810421.png">
*Safari*
<img width="336" alt="safari" src="https://user-images.githubusercontent.com/42325924/90692531-9231a700-e275-11ea-8224-39e99d2a240e.png">

We can notice that Google Chrome handles this differently.

**Comment by rsms** [comment]
Excellent news. Thank you @aemi-dev!

**Comment by rsms** [comment]
This has now been released as part of [version 3.15](https://github.com/rsms/inter/releases/tag/v3.15) 🎉
Try it out here: https://rsms.me/inter/lab/

**Comment by rsms** [comment]
Some ClearType rendering samples from Windows 10 using the v3.15 hinted font files:

<img width="888" alt="cleartype-samples@2x" src="https://user-images.githubusercontent.com/47046/90820678-9a75fa80-e2e6-11ea-912b-a40ebf3854fd.png">

**Comment by citizenkade** [comment]
Sorry to dredge up this issue. Just experienced it with Inter in Figma on a mac. Not a big deal. Just thought it was an interesting problem and Googled it. Ended up here.

![Screenshot 2024-11-12 at 3 00 57 PM](https://github.com/user-attachments/assets/a0a900a3-ae81-4c1a-a5cd-e82c734fa1df)
